manager/deallocator/deallocator.go

```diff
@@ -0,0 +1,244 @@
+package deallocator
+
+import (
+	"reflect"
+
+	"github.com/docker/go-events"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/docker/swarmkit/manager/state/store"
+	"golang.org/x/net/context"
+)
+
+var (
+	resourceDeallocators = make(map[reflect.Type]resourceDeallocator)
+)
+
+func registerServiceLevelResource(resourceDeallocator resourceDeallocator) {
+	resourceDeallocators[reflect.TypeOf(resourceDeallocator.event)] = resourceDeallocator
+}
+
+// the Deallocator waits for services to fully shutdown (ie no containers left)
+// and then proceeds to deallocate service-level resources (e.g. networks),
+// and finally services themselves
+type Deallocator struct {
+	store *store.MemoryStore
+
+	// for services that are shutting down, we keep track of how many
+	// tasks still exist for them
+	services map[string]serviceWithTaskCounts
+
+	stopChan chan struct{}
+	doneChan chan struct{}
+}
+
+type serviceWithTaskCounts struct {
+	service   *api.Service
+	taskCount int
+}
+
+// New creates a new deallocator
+func New(store *store.MemoryStore) *Deallocator {
+	return &Deallocator{
+		store:    store,
+		services: make(map[string]serviceWithTaskCounts),
+
+		stopChan: make(chan struct{}),
+		doneChan: make(chan struct{}),
+	}
+}
+
+// Run starts the deallocator, which then starts cleaning up services
+// and their resources when relevant (ie when no tasks still exist
+// for a given service)
+// This is a blocking function
+func (deallocator *Deallocator) Run(ctx context.Context) error {
+	// those are the events we'll be watching
+	events := make([]api.Event, len(resourceDeallocators)+2)
+	// we want to know when tasks are deleted by the reaper
+	events[0] = api.EventDeleteTask{}
+	// and when services get marked for deletion
+	events[1] = api.EventUpdateService{Service: &api.Service{PendingDelete: true}}
+	i := 2
+	for _, resourceDeallocator := range resourceDeallocators {
+		events[i] = resourceDeallocator.event
+		i++
+	}
+
+	eventsChan, eventsChanCancel, err := store.ViewAndWatch(deallocator.store,
+		func(readTx store.ReadTx) error {
+			// look for services that are marked for deletion
+			// there's no index on the `PendingDelete` field,
```

Comment:
>there's no index on the `PendingDelete` field

can you plz clarify what this means?

Reply:
I mean that there's no DB index on that field; so listing those which are marked for deletion requires iterating over them all.

Reply:
cool. maybe clarify then: 
 there's no index on the `PendingDelete` field _in the store_